spec/01-unit/01-db/01-schema/02-metaschema_spec.lua

```diff
@@ -501,3 +502,219 @@ describe("metaschema", function()
+      name = "bad",
+      fields = {
+        { foo = "bar", },
+      },
+      primary_key = { "foo" },
+    }
+    assert.falsy(MetaSchema.MetaSubSchema:validate(s))
+  end)
+
+  it("fields cannot be empty", function()
+    local s = {
+      name = "bad",
+      fields = {
+        {}
+      },
+    }
+    local ok, err = MetaSchema.MetaSubSchema:validate(s)
+    assert.falsy(ok)
+    assert.match("field entry table is empty", err.fields)
+  end)
+
+  it("rejects an invalid entity check", function()
+    local s = {
+      name = "bad",
+      fields = {
+        { foo = { type = "number" }, },
+      },
+      primary_key = { "foo" },
+      entity_checks = {
+        foo = { "bar" },
+      }
+    }
+    assert.falsy(MetaSchema.MetaSubSchema:validate(s))
+  end)
+
+  it("validates a schema with nested records", function()
+    local s = {
+      name = "hello",
+      fields = {
+        { foo = { type = "number" } },
+        { f = {
+            type = "record",
+            fields = {
+              { r = {
+                  type = "record",
+                  fields = {
+                    { a = { type = "string" }, },
+                    { b = { type = "number" }, } }}}}}}}}
+    assert.truthy(MetaSchema.MetaSubSchema:validate(s))
+  end)
+
+  it("allows only one entity check per array field", function()
+    local s = {
+      name = "bad",
+      fields = {
+        { a = { type = "number" } },
+        { b = { type = "number" } },
+        { c = { type = "number" } },
+        { d = { type = "number" } },
+      },
+      primary_key = { "foo" },
+      entity_checks = {
+        { only_one_of = { "a", "b" },
+          at_least_one_of = { "c", "d" },
+        },
+      }
+    }
+    local ok, errs = MetaSchema.MetaSubSchema:validate(s)
+    assert.falsy(ok)
+    assert.truthy(errs)
+  end)
+
+  it("accepts a function in an entity check", function()
+    local s = {
+      name = "bad",
+      fields = {
+        { a = { type = "number" } },
+        { b = { type = "number" } },
+      },
+      entity_checks = {
+        { custom_entity_check = {
+            field_sources = { "a" },
+            fn = function()
+              return true
+            end,
+          }
+        },
+      }
+    }
+    local ok = MetaSchema.MetaSubSchema:validate(s)
+    assert.truthy(ok)
+  end)
+
+  it("rejects a bad schema checking nested error", function()
+    local s = {
+      name = "bad",
+      fields = {
+        {
+          foo = {
+            type = "array",
+            elements = {
+              { foo = "bar", },
+            }
+          }
+        }
+      },
+      primary_key = { "foo" },
+    }
+    assert.falsy(MetaSchema.MetaSubSchema:validate(s))
+  end)
+
+  it("rejects a bad schema matching validators and types", function()
+    local s = {
+      name = "bad",
+      fields = {
+        {
+          foo = {
+            type = "array",
+            -- will cause error because `uuid` must be used with `strings`
```

Comment:
Transform this comment into an assert over `errs` at the end.

Reply:
Not needed given the updated error check.